ocaml/lambda/translcore.ml

```diff
@@ -216,57 +222,72 @@ let rec trivial_pat pat =
-        c_rhs={exp with exp_desc =
-                          Texp_function { arg_label; param; cases; partial;
-                                          region; curry; warnings; arg_mode;
-                                          arg_sort; ret_sort; alloc_mode }}}]
-  | [{c_lhs=pat; c_guard=None;
-      c_rhs={exp_attributes=[{Parsetree.attr_name = {txt="#default"};_}];
-             exp_desc = Texp_let
-               (Nonrecursive, binds,
-                ({exp_desc = Texp_function _} as e2))}}] ->
+      let exp_desc =
+        Texp_function
+          { arg_label; param; cases; partial; region; curry; warnings; arg_mode;
+            arg_sort; ret_sort; alloc_mode }
+      in
+      [ { c_lhs = pat; c_rhs = Simple_rhs { exp with exp_desc } } ]
+  | [{c_lhs=pat;
+      c_rhs=
+        Simple_rhs
+          { exp_attributes=[{Parsetree.attr_name = {txt="#default"};_}]
+          ; exp_desc = 
+              Texp_let
+                (Nonrecursive, binds, ({exp_desc = Texp_function _} as e2))}}
+    ] ->
       push_defaults loc (binds :: bindings) true
-                   arg_mode arg_sort [{c_lhs=pat;c_guard=None;c_rhs=e2}]
+                   arg_mode arg_sort [{c_lhs=pat;c_rhs=Simple_rhs e2}]
                    partial warnings
-  | [{c_lhs=pat; c_guard=None; c_rhs=exp} as case]
+  | [{c_lhs=pat; c_rhs=Simple_rhs exp} as case]
     when use_lhs || trivial_pat pat && exp.exp_desc <> Texp_unreachable ->
-      [{case with c_rhs = wrap_bindings bindings exp}]
-  | {c_lhs=pat; c_rhs=exp; c_guard=_} :: _ when bindings <> [] ->
+      [{case with c_rhs = Simple_rhs (wrap_bindings bindings exp)}]
+  | {c_lhs=pat; c_rhs=rhs } :: _
+    when bindings <> [] ->
+      let exp_loc, exp_extra, exp_type, exp_env, exp_attributes =
+        match rhs with 
+        | Simple_rhs e | Boolean_guarded_rhs { bg_rhs = e; _ } ->
+            e.exp_loc, e.exp_extra, e.exp_type, e.exp_env, e.exp_attributes
+        | Pattern_guarded_rhs { pg_loc; pg_env; pg_type; _ } ->
+            pg_loc, [], pg_type, pg_env, []
+      in
       let mode = Value_mode.of_alloc arg_mode in
       let param = Typecore.name_cases "param" cases in
       let desc =
         {val_type = pat.pat_type; val_kind = Val_reg;
          val_attributes = []; Types.val_loc = Location.none;
          val_uid = Types.Uid.internal_not_actually_unique; }
       in
-      let env = Env.add_value ~mode param desc exp.exp_env in
+      let env = Env.add_value ~mode param desc exp_env in
       let name = Ident.name param in
       let exp =
         let cases =
           let pure_case ({c_lhs; _} as case) =
             {case with c_lhs = as_computation_pattern c_lhs} in
           List.map pure_case cases in
-        { exp with exp_loc = loc; exp_env = env; exp_desc =
-          Texp_match
-            ({exp with exp_type = pat.pat_type; exp_env = env; exp_desc =
-              Texp_ident
+        { exp_loc = loc; exp_env = env; exp_extra; exp_type; exp_attributes;
```

Comment:
`exp_attributes` and `exp_extra` are doubly duplicated here. (They're being moved onto both the ident and the match expression, while also being preserved on the first case in the match's `cases`.) I really think this is a bug in the old code, and that these lists should be empty in the ident/match expression.

I'm not too concerned because `push_defaults` is getting deleted soon anyway. But we should leave a short comment explaining this weirdness, as it did temporarily take me aback.

Reply:
Got it, added a short comment explaining this